### DISCUSSION TITLE
/usr/local/bin/dialog as a POSIX shell script with bashisms

### DISCUSSION THREAD

**Discussion by fraserhess** [swiftDialog/swiftDialog#391] [General]
The `dialog` script has a shebang of `#!/bin/sh`. For now, scripts with that shebang are interpreted by Bash in sh compatibility mode by default.

However, the script uses bashisms such as arrays and `[[ ]]` tests that are not pure POSIX. This works because Bash is the interpreter. This situation should be examined as a matter of future-proofing.

As I see it, there are three scenarios:
* Apple switches to zsh for sh interpretation by default
* Apple switches to dash for sh interpretation by default
* A user or administrator using swiftDialog configures a Mac to use either of these options

Either Apple or a user switching to zsh is likely safe, but all permutations of the script should be tested.

If Apple or a user switched to dash, the script doesn't fail but has subtle errors.

#### Short term

Assuming macOS 15 doesn't drop Bash, `/usr/local/bin/dialog` should adopt `#!/bin/bash` in the short term. This keeps Bash as the interpreter but makes it explicit. Use of the `[ ]` POSIX test should be updated to use `[[ ]]`.

This could likely be merged in a bug-fix release.

#### Long term

`/usr/local/bin/dialog` should adopt `#!/bin/zsh`.  Use of the `[ ]` POSIX test should be updated to use `[[ ]]`.

This should wait for a 2.6 or later release unless macOS 15 drops Bash, making this concern an emergency.

----
I am happy to create pull requests if these changes are desired.

Reference: https://scriptingosx.com/2020/06/about-bash-zsh-sh-and-dash-in-macos-catalina-and-beyond/

**Comment by bartreardon** [comment]
yeah, the script should agree with whatever context it claims to be running in. 

I have a preference for zsh but also like to make things as "lowest common denominator" as possible for things like this and so would probably defer to `sh` wherever that ends up being redirected to. I don't think using zsh would be an issue though since we're not relying on any advanced functionality.

day to day I write all my stuff in zsh
